crates/biome_ruledoc_utils/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+mod codeblock;
+
+use std::collections::HashMap;
+use std::hash::BuildHasher;
+use std::sync::Arc;
+
+use biome_fs::{BiomePath, MemoryFileSystem};
+use biome_js_analyze::JsAnalyzerServices;
+use biome_js_parser::JsFileSource;
+use biome_json_parser::{JsonParserOptions, parse_json};
+use biome_module_graph::ModuleGraph;
+use biome_project_layout::ProjectLayout;
+use biome_test_utils::get_added_paths;
+use camino::Utf8PathBuf;
+
+pub use codeblock::*;
+
+/// Builder that can be used for constructing analyzer services.
+///
+/// The builder can be reused to create cheap instances of analyzer services
+/// for multiple code blocks.
+pub struct AnalyzerServicesBuilder {
```

Comment:
I refactored a little bit how we create services for code block analysis. Previously, this was done ad-hoc, which resulted in indexing of the module graph and project layout for every block separately. Now, when multi-file rules are analysed, we index once and reuse the module graph and project layout between code blocks.